### PR TITLE
Initial work on running windows containers on Kubernetes

### LINKED ISSUE #30279: Kubelet support on Windows
Supporting the kubelet on Windows is one of the first steps to getting Kubernetes working on Windows. The scope of this issue is to get the kubelet up and running on Windows, and to have the ability to deploy a pod that has a windows-based container on a Windows node.

### PR DESCRIPTION
**Pull request by alexbrand**
This is the first stab at getting the Kubelet running on Windows (fixes #30279), and getting it to deploy network-accessible pods that consist of Windows containers. Thanks @csrwng, @jbhurat for helping out.

The main challenge with Windows containers at this point is that container networking is not supported. In other words, each container in the pod will get it's own IP address. For this reason, we had to make a couple of changes to the kubelet when it comes to setting the pod's IP in the Pod Status. Instead of using the infra-container's IP, we use the IP address of the first container.

Other approaches we investigated involved "disabling" the infra container, either conditionally on `runtime.GOOS` or having a separate windows-docker container runtime that re-implemented some of the methods (would require some refactoring to avoid maintainability nightmare). 

Other changes:
- The default docker endpoint was removed. This results in the docker client using the default for the specific underlying OS.

More detailed documentation on how to setup the Windows kubelet can be found at https://docs.google.com/document/d/1IjwqpwuRdwcuWXuPSxP-uIz0eoJNfAJ9MWwfY20uH3Q. 

cc: @ikester @brendandburns @jstarks

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31707)

### CODE REVIEW

**Comment by brendandburns** [comment]
@k8s-bot ok to test

@jbhurat can you validate that you have signed the CLA?  I know that Alex and Caesar have since I can see their commit history...

File: pkg/apis/componentconfig/v1alpha1/defaults.go [hunk 1]
```diff
@@ -176,9 +176,6 @@ func SetDefaults_KubeletConfiguration(obj *KubeletConfiguration) {
 	if obj.DockerExecHandlerName == "" {
 		obj.DockerExecHandlerName = "native"
 	}
-	if obj.DockerEndpoint == "" {
```
Comment (brendandburns) [inline pkg/apis/componentconfig/v1alpha1/defaults.go]: It seems like a better answer is conditionalize this on !windows

File: pkg/kubelet/dockertools/docker_manager.go [hunk 2]
```diff
@@ -410,13 +401,11 @@ func (dm *DockerManager) inspectContainer(id string, podName, podNamespace strin
 		// Container that are running, restarting and paused
 		status.State = kubecontainer.ContainerStateRunning
 		status.StartedAt = startedAt
-		if containerName == PodInfraContainerName {
```
Comment (brendandburns) [inline pkg/kubelet/dockertools/docker_manager.go]: Why did you remove this check?

File: pkg/volume/volume.go [hunk 3]
```diff
@@ -176,9 +177,63 @@ func RenameDirectory(oldPath, newName string) (string, error) {
 	if err != nil {
 		return "", err
 	}
-	err = os.Rename(oldPath, newPath)
+	// os.Rename call fails on windows (https://github.com/golang/go/issues/14527)
```
Comment (brendandburns) [inline pkg/volume/volume.go]: Let's conditionalize this on runtime.GOOS==windows

**Comment by jbhurat** [comment]
I signed it! (Covered under Corporate Agreement - Apprenda)

File: pkg/kubelet/dockertools/docker_manager.go (hunk 2 above)
Comment (alexbrand) [inline pkg/kubelet/dockertools/docker_manager.go]: My thinking was that `inspectContainer` should return the IP address for any given container. We then determine the pod IP in `GetPodStatus`, using the `containerProvidesPodIP` func to determine which container's IP we should use. On the Linux side, we want the infra container's IP. On the windows side, we don't want the infra container's IP.

**Comment by csrwng** [comment]
@alexbrand I agree my commits can be used. Not sure how to tell that to k8s-bot.

**Comment by alexbrand** [comment]
Looks like the build ran into https://github.com/kubernetes/kubernetes/issues/27462 here.

**Comment by brendandburns** [comment]
I manually marked CLA human approved.

But this needs a rebase, etc.

**Comment by brendandburns** [comment]
@alexbrand can you look into the failing tests?

**Comment by alexbrand** [comment]
@brendandburns Certainly, looking at them now.

**Comment by pires** [comment]
@brendandburns is it needed for both CLAs, Google and CNCF? 🤔

File: pkg/api/validation/validation.go [hunk 4]
```diff
@@ -1463,8 +1463,6 @@ func validateVolumeMounts(mounts []api.VolumeMount, volumes sets.String, fldPath
 		}
 		if len(mnt.MountPath) == 0 {
 			allErrs = append(allErrs, field.Required(idxPath.Child("mountPath"), ""))
-		} else if strings.Contains(mnt.MountPath, ":") {
```
Comment (brendandburns) [inline pkg/api/validation/validation.go]: doesn't seem like this should be removed?

Comment (brendandburns) [inline pkg/api/validation/validation.go]: or at least conditionalize this on windows?

File: pkg/kubelet/dockertools/docker_manager.go [hunk 5]
```diff
@@ -1122,23 +1120,6 @@ func (dm *DockerManager) fmtDockerOpts(opts []dockerOpt) ([]string, error) {
 	return fmtOpts, nil
 }
 
-func (dm *DockerManager) getSecurityOpts(pod *api.Pod, ctrName string) ([]dockerOpt, error) {
```
Comment (brendandburns) [inline pkg/kubelet/dockertools/docker_manager.go]: Can you restore this function to this location so that we minimize the diff?

Comment (brendandburns) [inline pkg/kubelet/dockertools/docker_manager.go]: (ah, nm you moved this out to a different file...)

**Comment by brendandburns** [comment]
This LGTM, with some small comments.

File: pkg/kubelet/dockertools/docker_manager.go (hunk 5 above)
Comment (pires) [inline pkg/kubelet/dockertools/docker_manager.go]: Yeah, needs different impls for now.

File: pkg/api/validation/validation.go (hunk 4 above)
Comment (pires) [inline pkg/api/validation/validation.go]: My question here is, where exactly does this runs, the kubelet? I was under the impression this validation would happen at the API server and having that said, just couldn't conditionalize on Windows, e.g. `if GOOS=="windows"`.

Comment (pires) [inline pkg/api/validation/validation.go]: On it.

Comment (pires) [inline pkg/api/validation/validation.go]: Actually, it seems that:
* `c:/my/path` works with Windows Containers (there's some translation in place)
* `/my:path` is valid Unix path

Please, advise.

**Comment by brendandburns** [comment]
LGTM

**Comment by pires** [comment]
@brendandburns can you label this to be released with 1.5?
